File: io/lucamienert/basic/ast/PrintStatement.java
package io.lucamienert.basic.ast;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Collections;

import io.lucamienert.basic.instruction.Instruction;
import io.lucamienert.basic.instruction.InstructionSequence;
import io.lucamienert.basic.instruction.OpCode;

public class PrintStatement extends Statement {

    private final List<StringExpression> values;

    public PrintStatement(StringExpression... values) {
        this.values = Collections.unmodifiableList(Arrays.asList(values));
    }

    public PrintStatement(List<StringExpression> values) {
        this.values = Collections.unmodifiableList(new ArrayList<>(values));
    }

    public List<StringExpression> getValues() {
        return values;
    }

    @Override
    public void compile(InstructionSequence seq) {
        for (StringExpression value : values) {
            value.compile(seq);

            seq.append(new Instruction(value instanceof ImmediateString ? OpCode.OUTS : OpCode.OUTI));
        }

        seq.append(new Instruction(OpCode.PUSHS, "\n"), new Instruction(OpCode.OUTS));
    }
    
}

File: BasicCompiler/Ast/PrintStatement.cs
using BasicCompiler.Instructions;

namespace BasicCompiler.Ast;

public class PrintStatement : Statement
{
    public List<StringExpression> Values { get; set; }

    public PrintStatement(params StringExpression[] values)
        => Values = new List<StringExpression>(values);

    public override void Compile(InstructionSequence instructionSequence)
    {
        foreach(var value in Values)
        {
            value.Compile(instructionSequence);
            instructionSequence.Append(new Instruction(value is ImmediateString ? OpCode.OUTS : OpCode.OUTI));
        }

        instructionSequence.Append(
            new Instruction(OpCode.PUSHS, "\n"), 
            new Instruction(OpCode.OUTS)
        );
    }
}
